File: javascript/validate_tool_output.js
#!/usr/bin/env node
// validate_tool_output.py — portiert nach javascript
// Quelle: python, OpenClaw@gateway1:skills/json-utils/scripts/validate_tool_output.py
// auch in: OpenClaw@gateway2:skills/json-utils/scripts/validate_tool_output.py
// Erzeugt: 2026-08-22 durch ABSTRACTIONS_MANAGER.py

/**
 * Validiert Tool-Outputs gegen ein Pydantic-Schema.
 * Für OpenClaw Tool-Call-Validierung.
 */

const fs = require('fs');
const path = require('path');
const process = require('process');

// Zod als Ersatz für Pydantic
const { z } = require('zod');

class JSONValidationError extends Error {
  constructor(message) {
    super(message);
    this.name = 'JSONValidationError';
  }
}

function parseAndValidate(jsonInput, schema, options = {}) {
  const { repair = true, strict = false } = options;
  
  let jsonData;
  try {
    jsonData = typeof jsonInput === 'string' ? JSON.parse(jsonInput) : jsonInput;
  } catch (error) {
    throw new JSONValidationError(`Invalid JSON: ${error.message}`);
  }
  
  try {
    const result = schema.parse(jsonData);
    return result;
  } catch (error) {
    if (repair && error instanceof z.ZodError) {
      // Versuche Reparatur durch partielle Validierung
      try {
        const partialResult = {};
        for (const key in jsonData) {
          try {
            if (schema.shape[key]) {
              partialResult[key] = schema.shape[key].parse(jsonData[key]);
            } else if (!strict) {
              partialResult[key] = jsonData[key];
            }
          } catch (fieldError) {
            if (!strict) {
              partialResult[key] = jsonData[key];
            }
          }
        }
        return partialResult;
      } catch (partialError) {
        throw new JSONValidationError(`Validation failed: ${error.message}`);
      }
    }
    throw new JSONValidationError(`Validation failed: ${error.message}`);
  }
}

function createDynamicModel(schema) {
  /** Erstellt ein dynamisches Zod-Modell aus einem JSON-Schema. */
  const fields = {};
  
  const properties = schema.properties || {};
  const requiredFields = new Set(schema.required || []);
  
  for (const [fieldName, fieldInfo] of Object.entries(properties)) {
    let fieldSchema;
    
    const jsonType = fieldInfo.type || 'string';
    switch (jsonType) {
      case 'integer':
        fieldSchema = z.number().int();
        break;
      case 'number':
        fieldSchema = z.number();
        break;
      case 'boolean':
        fieldSchema = z.boolean();
        break;
      case 'array':
        fieldSchema = z.array(z.any());
        break;
      case 'object':
        fieldSchema = z.record(z.any());
        break;
      case 'string':
      default:
        fieldSchema = z.string();
    }
    
    // Beschreibung hinzufügen, falls vorhanden
    if (fieldInfo.description) {
      fieldSchema = fieldSchema.describe(fieldInfo.description);
    }
    
    // Default-Wert setzen
    if (fieldInfo.default !== undefined) {
      fieldSchema = fieldSchema.default(fieldInfo.default);
    }
    
    // Feld optional machen, wenn nicht required
    if (!requiredFields.has(fieldName)) {
      fieldSchema = fieldSchema.optional();
      if (fieldInfo.default === undefined) {
        fieldSchema = fieldSchema.nullable();
      }
    }
    
    fields[fieldName] = fieldSchema;
  }
  
  return z.object(fields);
}

function main() {
  const args = parseArguments();
  
  // Lade Schema
  let schema;
  try {
    const schemaContent = fs.readFileSync(args.schema, 'utf8');
    schema = JSON.parse(schemaContent);
  } catch (error) {
    console.error(`Error loading schema: ${error.message}`);
    process.exit(1);
  }
  
  // Lade Input
  let rawInput;
  try {
    if (args.file) {
      rawInput = fs.readFileSync(args.json_input, 'utf8');
    } else {
      rawInput = args.json_input;
    }
  } catch (error) {
    console.error(`Error loading input: ${error.message}`);
    process.exit(1);
  }
  
  // Erstelle dynamisches Modell und validiere
  try {
    const modelSchema = createDynamicModel(schema);
    const result = parseAndValidate(rawInput, modelSchema, { 
      repair: args.repair, 
      strict: args.strict 
    });
    console.log(JSON.stringify(result, null, 2));
  } catch (error) {
    if (error instanceof JSONValidationError) {
      console.error(`Validation error: ${error.message}`);
    } else {
      console.error(`Unexpected error: ${error.message}`);
    }
    process.exit(1);
  }
}

function parseArguments() {
  const argv = process.argv.slice(2);
  const args = {
    json_input: null,
    schema: null,
    file: false,
    repair: true,
    strict: false
  };
  
  const positionalArgs = [];
  
  for (let i = 0; i < argv.length; i++) {
    const arg = argv[i];
    
    if (arg === '--schema' || arg === '-s') {
      args.schema = argv[++i];
    } else if (arg === '--file' || arg === '-f') {
      args.file = true;
    } else if (arg === '--repair' || arg === '-r') {
      args.repair = true;
    } else if (arg === '--no-repair') {
      args.repair = false;
    } else if (arg === '--strict') {
      args.strict = true;
    } else if (!arg.startsWith('-')) {
      positionalArgs.push(arg);
    }
  }
  
  if (positionalArgs.length > 0) {
    args.json_input = positionalArgs[0];
  }
  
  if (!args.json_input) {
    console.error('Error: json_input argument is required');
    process.exit(1);
  }
  
  if (!args.schema) {
    console.error('Error: --schema/-s option is required');
    process.exit(1);
  }
  
  return args;
}

if (require.main === module) {
  main();
}

module.exports = {
  createDynamicModel,
  parseAndValidate,
  JSONValidationError
};

File: perl5/validate_tool_output.pl
#!/usr/bin/perl
# validate_tool_output.py — portiert nach perl5
# Quelle: python, OpenClaw@gateway1:skills/json-utils/scripts/validate_tool_output.py
# auch in: OpenClaw@gateway2:skills/json-utils/scripts/validate_tool_output.py
# Erzeugt: 2026-08-22 durch ABSTRACTIONS_MANAGER.py

use strict;
use warnings;
use Getopt::Long;
use JSON;
use File::Slurp;
use Data::Dumper;

# Hilfsfunktionen für Fehlerausgabe
sub error {
    my ($msg) = @_;
    print STDERR "Error: $msg\n";
    exit 1;
}

# Dynamisches Modell erstellen (vereinfacht)
sub create_dynamic_model {
    my ($schema) = @_;
    my %model;
    
    my $properties = $schema->{properties} || {};
    my $required = { map { $_ => 1 } @{$schema->{required} || []} };
    
    for my $field_name (keys %$properties) {
        my $field_info = $properties->{$field_name};
        my $field_type = 'string'; # Default
        
        my $json_type = $field_info->{type} || 'string';
        if ($json_type eq 'integer') {
            $field_type = 'integer';
        } elsif ($json_type eq 'number') {
            $field_type = 'number';
        } elsif ($json_type eq 'boolean') {
            $field_type = 'boolean';
        } elsif ($json_type eq 'array') {
            $field_type = 'array';
        } elsif ($json_type eq 'object') {
            $field_type = 'object';
        }
        
        my $default = exists $field_info->{default} ? $field_info->{default} : undef;
        unless ($required->{$field_name}) {
            $default = exists $field_info->{default} ? $field_info->{default} : undef;
        }
        
        $model{$field_name} = {
            type => $field_type,
            default => $default,
            description => $field_info->{description} || '',
        };
    }
    
    return \%model;
}

# Validierungsfunktion (vereinfacht)
sub parse_and_validate {
    my ($raw_input, $model, $repair, $strict) = @_;
    
    my $data;
    eval {
        $data = decode_json($raw_input);
    };
    if ($@) {
        error("Invalid JSON input: $@");
    }
    
    my $result = {};
    for my $field_name (keys %$model) {
        my $field_def = $model->{$field_name};
        my $value = $data->{$field_name};
        
        if (!defined $value && defined $field_def->{default}) {
            $value = $field_def->{default};
        }
        
        if (!defined $value) {
            next unless $strict;
            error("Missing required field: $field_name");
        }
        
        # Typvalidierung (vereinfacht)
        if ($field_def->{type} eq 'integer' && defined $value) {
            if (ref $value || $value !~ /^-?\d+$/) {
                error("Field '$field_name' must be an integer") unless $repair;
                $value = int($value);
            }
        } elsif ($field_def->{type} eq 'number' && defined $value) {
            if (ref $value || $value !~ /^-?\d+(\.\d+)?$/) {
                error("Field '$field_name' must be a number") unless $repair;
                $value = $value + 0;
            }
        } elsif ($field_def->{type} eq 'boolean' && defined $value) {
            if ($value ne 'true' && $value ne 'false' && $value != 0 && $value != 1) {
                error("Field '$field_name' must be a boolean") unless $repair;
                $value = $value ? 'true' : 'false';
            }
        } elsif ($field_def->{type} eq 'array' && defined $value) {
            if (ref $value ne 'ARRAY') {
                error("Field '$field_name' must be an array") unless $repair;
                $value = [$value];
            }
        } elsif ($field_def->{type} eq 'object' && defined $value) {
            if (ref $value ne 'HASH') {
                error("Field '$field_name' must be an object") unless $repair;
                $value = { value => $value };
            }
        }
        
        $result->{$field_name} = $value;
    }
    
    return $result;
}

# Hauptprogramm
sub main {
    my $json_input;
    my $schema_file;
    my $is_file = 0;
    my $repair = 1;
    my $strict = 0;
    
    GetOptions(
        'schema|s=s' => \$schema_file,
        'file|f'     => \$is_file,
        'repair|r'   => \$repair,
        'strict'     => \$strict,
    ) or error("Invalid options");
    
    @ARGV or error("Missing json_input argument");
    $json_input = $ARGV[0];
    
    $schema_file or error("--schema is required");
    
    # Lade Schema
    my $schema_content;
    eval {
        $schema_content = read_file($schema_file);
    };
    if ($@) {
        error("Error loading schema: $@");
    }
    
    my $schema;
    eval {
        $schema = decode_json($schema_content);
    };
    if ($@) {
        error("Invalid JSON in schema file: $@");
    }
    
    # Lade Input
    my $raw_input;
    if ($is_file) {
        eval {
            $raw_input = read_file($json_input);
        };
        if ($@) {
            error("Error loading input file: $@");
        }
    } else {
        $raw_input = $json_input;
    }
    
    # Erstelle dynamisches Modell und validiere
    my $model = create_dynamic_model($schema);
    my $result;
    eval {
        $result = parse_and_validate($raw_input, $model, $repair, $strict);
    };
    if ($@) {
        error("Validation error: $@");
    }
    
    print encode_json($result) . "\n";
}

main() unless caller;
